File: apps/go/auth/test/integration/auth_test.go
//go:build integration

package integration_test

import (
	"encoding/json"
	"net/http"
	"testing"
	"time"

	"github.com/stretchr/testify/require"
)

// TestAuth covers public /api/auth/* endpoints.
func TestAuth(t *testing.T) {
	group(t, "POST /api/auth/register", func(t *testing.T) {
		t.Run("creates a user with valid payload and returns 201 + id", func(t *testing.T) {
			client := freshClient(t)
			email := randEmail()
			beforeEvents := redisStreamLen(t, "auth:user-events")
			step(t, "registering %s", email)

			response := postJSON(t, client, authBase+"/api/auth/register", map[string]string{
				"email": email, "password": "password123", "name": "Bob", "family": "B",
			})
			defer response.Body.Close()
			responseBody := readBody(response)
			require.Equalf(t, 201, response.StatusCode, "body: %s", responseBody)

			var body struct {
				ID             string `json:"id"`
				IsInitializing bool   `json:"isInitializing"`
			}
			require.NoError(t, json.Unmarshal([]byte(responseBody), &body))
			step(t, "got user id=%s", body.ID)
			require.NotEmpty(t, body.ID)
			require.True(t, body.IsInitializing)
			require.Equal(t, beforeEvents+1, redisStreamLen(t, "auth:user-events"))
		})

		t.Run("rejects duplicate email with 409", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")
			beforeEvents := redisStreamLen(t, "auth:user-events")

			step(t, "second register with same email must conflict")
			response := postJSON(t, client, authBase+"/api/auth/register", map[string]string{
				"email": email, "password": "password123", "name": "Bob", "family": "B",
			})
			defer response.Body.Close()
			require.Equal(t, 409, response.StatusCode)
			require.Equal(t, beforeEvents, redisStreamLen(t, "auth:user-events"))
		})

		t.Run("rejects invalid payload with 400", func(t *testing.T) {
			client := freshClient(t)
			beforeEvents := redisStreamLen(t, "auth:user-events")
			step(t, "sending bad email + short password")
			response := postJSON(t, client, authBase+"/api/auth/register", map[string]string{
				"email": "not-an-email", "password": "short", "name": "", "family": "",
			})
			defer response.Body.Close()
			require.Equal(t, 400, response.StatusCode)
			require.Equal(t, beforeEvents, redisStreamLen(t, "auth:user-events"))
		})
	})

	group(t, "POST /api/auth/login", func(t *testing.T) {
		t.Run("returns 204 + HttpOnly session cookie on success", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")

			step(t, "logging in %s", email)
			response := loginUser(t, client, email, "password123")
			defer response.Body.Close()
			require.Equal(t, 204, response.StatusCode)

			var sessionCookie *http.Cookie
			for _, cookie := range response.Cookies() {
				if cookie.Name == "session" {
					sessionCookie = cookie
				}
			}
			require.NotNil(t, sessionCookie, "session cookie must be set")
			step(t, "cookie value len=%d, HttpOnly=%v, MaxAge=%d", len(sessionCookie.Value), sessionCookie.HttpOnly, sessionCookie.MaxAge)
			require.NotEmpty(t, sessionCookie.Value)
			require.True(t, sessionCookie.HttpOnly)
		})

		t.Run("returns 401 on wrong password", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")

			response := loginUser(t, client, email, "totally-wrong")
			defer response.Body.Close()
			require.Equal(t, 401, response.StatusCode)
		})

		t.Run("locks account after N failed attempts and unlocks after duration", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")

			step(t, "spamming 3 wrong passwords")
			for attemptIndex := 0; attemptIndex < 3; attemptIndex++ {
				response := loginUser(t, client, email, "wrong-password")
				require.Equalf(t, 401, response.StatusCode, "attempt %d", attemptIndex+1)
				response.Body.Close()
			}

			step(t, "next attempt must be 423 Locked")
			response := loginUser(t, client, email, "wrong-password")
			require.Equal(t, 423, response.StatusCode)
			response.Body.Close()

			step(t, "waiting 3s for lockout to expire")
			time.Sleep(3 * time.Second)

			step(t, "correct password should succeed again")
			response = loginUser(t, client, email, "password123")
			defer response.Body.Close()
			require.Equalf(t, 204, response.StatusCode, "body: %s", readBody(response))
		})
	})

	group(t, "GET /api/auth/me and POST /api/auth/logout", func(t *testing.T) {
		t.Run("returns 401 without cookie", func(t *testing.T) {
			client := freshClient(t)
			response := get(t, client, authBase+"/api/auth/me")
			defer response.Body.Close()
			require.Equal(t, 401, response.StatusCode)
		})

		t.Run("returns profile with initialization flag after login", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")
			loginResponse := loginUser(t, client, email, "password123")
			loginResponse.Body.Close()

			response := get(t, client, authBase+"/api/auth/me")
			defer response.Body.Close()
			require.Equal(t, 200, response.StatusCode)

			var body struct {
				ID             string `json:"id"`
				Email          string `json:"email"`
				Name           string `json:"name"`
				Family         string `json:"family"`
				IsInitializing bool   `json:"isInitializing"`
				Organizations  []any  `json:"organizations"`
			}
			require.NoError(t, json.NewDecoder(response.Body).Decode(&body))
			step(t, "got id=%s email=%s orgs=%d", body.ID, body.Email, len(body.Organizations))
			require.Equal(t, email, body.Email)
			require.NotEmpty(t, body.ID)
			require.True(t, body.IsInitializing)
			require.Empty(t, body.Organizations, "auth must return an empty list before server initializes orgs")
		})

		t.Run("logout clears cookie and subsequent /me returns 401", func(t *testing.T) {
			client := freshClient(t)
			email := signUpUser(t, client, "password123")
			loginResponse := loginUser(t, client, email, "password123")
			loginResponse.Body.Close()

			step(t, "calling /logout")
			logoutResponse := postJSON(t, client, authBase+"/api/auth/logout", nil)
			require.Equal(t, 204, logoutResponse.StatusCode)
			logoutResponse.Body.Close()

			step(t, "/me must now be 401")
			response := get(t, client, authBase+"/api/auth/me")
			defer response.Body.Close()
			require.Equal(t, 401, response.StatusCode)
		})
	})
}

File: e2e/backend/scenarios/auth.test.ts
import { randomUUID } from 'node:crypto';
import { describe, expect, it } from 'vitest';

import { api, extractSessionCookie } from '../utils/api.js';
import { registerAndLogin } from '../utils/auth.js';

describe('auth: registration, login, session lifecycle', () => {
    it('register -> creates user, login returns a session cookie', async () => {
        const email = `auth-e2e-${randomUUID()}@example.com`;
        const password = 'S3curePassword!';

        const register = await api('/api/auth/register', {
            method: 'POST',
            body: JSON.stringify({ email, password, name: 'Auth', family: 'Test' }),
        });
        expect([201, 204]).toContain(register.status);

        const login = await api('/api/auth/login', {
            method: 'POST',
            body: JSON.stringify({ email, password }),
        });
        expect([200, 204]).toContain(login.status);

        const cookie = extractSessionCookie(login);
        expect(cookie).toMatch(/^session=/);
    });

    it('/me returns user with auto-created first organization', async () => {
        const user = await registerAndLogin();

        const response = await api('/api/auth/me', { cookie: user.cookie });
        expect(response.status).toBe(200);

        const me = (await response.json()) as {
            id: string;
            email: string;
            isInitializing: boolean;
            organizations: Array<{ id: string; role: string }>;
        };

        expect(me.id).toMatch(/^[0-9a-f-]{36}$/);
        expect(me.email).toBe(user.email);
        expect(me.isInitializing).toBe(false);
        expect(me.organizations.length).toBeGreaterThanOrEqual(1);
        expect(me.organizations.some(o => o.id === user.orgId && o.role === 'owner')).toBe(
            true
        );
    });

    it('/me without cookie -> 401', async () => {
        const response = await api('/api/auth/me');
        expect(response.status).toBe(401);
    });

    it('login with wrong password -> 401 and no session cookie', async () => {
        const user = await registerAndLogin();

        const response = await api('/api/auth/login', {
            method: 'POST',
            body: JSON.stringify({ email: user.email, password: 'wrong-password' }),
        });
        expect(response.status).toBe(401);
        expect(extractSessionCookie(response)).toBeNull();
    });

    it('login with unknown email -> 401', async () => {
        const response = await api('/api/auth/login', {
            method: 'POST',
            body: JSON.stringify({
                email: `nobody-${randomUUID()}@example.com`,
                password: 'whatever-1!',
            }),
        });
        expect(response.status).toBe(401);
    });

    it('logout invalidates the session', async () => {
        const user = await registerAndLogin();

        const logout = await api('/api/auth/logout', {
            method: 'POST',
            cookie: user.cookie,
        });
        expect([200, 204]).toContain(logout.status);

        const me = await api('/api/auth/me', { cookie: user.cookie });
        expect(me.status).toBe(401);
    });

    it('account is temporarily locked after too many failed attempts', async () => {
        const user = await registerAndLogin();

        let sawLockout = false;
        for (let i = 0; i < 7; i++) {
            const response = await api('/api/auth/login', {
                method: 'POST',
                body: JSON.stringify({
                    email: user.email,
                    password: 'definitely-wrong',
                }),
            });
            if (response.status === 423 || response.status === 429) {
                sawLockout = true;

                break;
            }
        }
        expect(sawLockout).toBe(true);
    });
});
